src/cargo/util/errors.rs

```diff
@@ -286,6 +286,16 @@ impl From<Box<CargoError>> for CliError {
 }
 
 // =============================================================================
+// NetworkError trait
+
+pub trait NetworkError: Error + Send + 'static {
+    fn maybe_spurious(&self) -> bool { true }
```

Comment:
Yeah this looks along the right lines to me, but you may want to define it as:

```rust
pub trait NetworkError: CargoError {
    fn maybe_spurious(&self) -> bool;
}
```

That way you can elide the extra bounds by just inheriting from `CargoError`, and it also forces the implementations below to check correctly whether they're related to the network before unconditionally retrying.